### PR TITLE
zh-cn: init the translation of "CycleTracker: Service workers"

### PR DESCRIPTION
**Pull request by BSPR0002**
### Motivation

files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md

### Related issues and pull requests

Relates to #15548

### CODE REVIEW

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+
+我们有了版本号，也有了需要缓存的文件。在缓存文件之前，我们需要创建一个在存储应用的静态资源时需要使用的缓存名称。这个缓存名称应当被版本化，以确保在应用更新时可以创建新的缓存同时删除旧的缓存。
+
+#### 任务
+
+使用 `VERSION` 版本号赖创建一个版本化的 `CACHE_NAME` 缓存名，作为常量添加到 `sw.js`。
+
+#### 示例方案
+
+我们将我们缓存命名为 `period-tracker-` 后跟 `VERSION`版本号。鉴于我们的常量声明都是单行的，为了方便阅读，我们将其放在了资源数组常量的前面。
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:106]:
```suggestion
我们将缓存命名为 `period-tracker-` 后跟 `VERSION`版本号。鉴于我们的常量声明都是单行的，为了方便阅读，我们将其放在了资源数组常量的前面。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+      } else {
+        // 如果资源不在缓存中，返回 404
+        return new Response(null, { status: 404 });
+      }
+    })()
+  );
+});
+```
+
+当更新 service worker 本身时，不需要更新 VERSION 常量，因为 service worker 脚本内容的任何更改都会触发浏览器去安装新的 service worker。但不论如何，更新版本号是方便开发人员，也包括你自己的好习惯。可以在浏览器中通过[在应用程序工具中检查缓存的名称](#with_developer_tools)（或在源代码工具）来查看当前正在运行的是哪一个版本的 service worker。
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:301]:
```suggestion
当更新 service worker 本身时，不需要更新 VERSION 常量，因为 service worker 脚本内容的任何更改都会触发浏览器去安装新的 service worker。但不论如何，更新版本号是方便开发人员的好习惯，也包括你自己。可以在浏览器中通过[在应用程序工具中检查缓存的名称](#with_developer_tools)（或在源代码工具）来查看当前正在运行的是哪一个版本的 service worker。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+```html
+<!-- 注册应用的 service worker -->
+<script>
+  if ("serviceWorker" in navigator) {
+    navigator.serviceWorker.register("sw.js");
+  }
+</script>
+```
+
+您可以体验功能完整的[CycleTracker 月经周期跟踪 Web 应用](https://mdn.github.io/pwa-examples/cycletracker/service_workers)，并可以在 GitHub 上查看该 [Web 应用的源代码](https://github.com/mdn/pwa-examples/tree/master/cycletracker/service_workers)。是的，它能运行，并且现在正式成为了一个 PWA！
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:361]:
```suggestion
你可以体验功能完整的[CycleTracker 月经周期跟踪 Web 应用](https://mdn.github.io/pwa-examples/cycletracker/service_workers)，并可以在 GitHub 上查看该 [Web 应用的源代码](https://github.com/mdn/pwa-examples/tree/master/cycletracker/service_workers)。是的，它能运行，并且现在正式成为了一个 PWA！
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+出于我们安装 service worker 的方式的因素，一旦其被注册，所有的请求都将会从缓存中拉取，而不是加载新内容。在开发的时候，你会频繁地编辑你的代码，你可能会希望定期甚至是在每一次保存后在浏览器中测试你的编辑。
+
+### 通过更新版本号并进行强重置
+
+要获得新的缓存，你可以更改[版本号](#版本号)然后进行一次浏览器强刷新。进行强刷新的方式取决于你的浏览器和操作系统：
+
+- Windows：Ctrl+F5，Shift+F5 或 Ctrl+Shift+R。
+- MacOS：Shift+Command+R。
+- MacOS 上的 Safari：按 Option+Command+E 清空缓存，然后按 Option+Command+R。
+- 移动设备：前往浏览器（Android）或操作系统（Samsung、iOS）设置，在高级设置下找到浏览器（iOS）或站点数据（Android、Samsung）网站设置，然后删除经期跟踪器的数据，再重新下载页面。
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:375]:
```suggestion
- 移动设备：前往浏览器（Android）或操作系统（Samsung、iOS）设置，在高级设置下找到浏览器（iOS）或站点数据（Android、Samsung）网站设置，然后删除经期跟踪器的数据，再重新加载页面。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+
+如果你还没有完成到这一步，请复制这些 [HTML](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/index.html)、[CSS](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/style.css)、[JavaScript](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/app.js)，还有[清单](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/cycletracker.json) JSON 文件的内容，并分别按照以下文件名保存：`index.html`、`styles.css`、`app.js`，以及 `cycletracker.json`。
+
+在本章节，我们要创建 `sw.js`——service worker 脚本，用于将我们的 Web 应用转化成 PWA。我们已经有了一个 JavaScript 文件——在 HTML 文件的最后一行里名为 `app.js`的那一个。这个 JavaScript 脚本提供了全部标准 Web 应用特性的功能。我们并不会像使用 `app.js` 那样通过 {{HTMLElement("script")}} 元素的 `src` 属性来调用 `sw.js`，而是通过注册 service worker 来建立它与 Web 应用的联系。
+
+在本课程的最后，你将会拥有一个功能完整的 PWA——一个能够被完整安装的、即使用户处于离线状态也能运行的逐步增强的 Web 应用。
+
+## Service worker 的职责
+
+Serbice worker 能够让应用离线运行，并且时刻确保应用是最新的。为了能出色地做到这点，service worker 应当包含以下信息：
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:21]:
```suggestion
Service worker 能够让应用离线运行，并且时刻确保应用是最新的。为了能出色地做到这点，service worker 应当包含以下信息：
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+
+将上述 `activate` 事件监听器添加至你的 `sw.js` 文件。
+
+### fetch 事件
+
+我们可以利用 [`fetch`](/zh-CN/docs/Web/API/ServiceWorkerGlobalScope/fetch_event) 事件的优势，在用户处于在线状态时阻止已安装的 PWA 发起请求。监听 fetch 事件使得拦截所有请求并使用缓存的响应内容进行响应而不流经网络变得可能。大多数的应用不需要这样的行为。事实上，许多商业模式出于跟踪和营销目的都希望用户定期发送服务器请求。所以，尽管拦截请求对于某些应用来说是一种反面模式，但为了提高我们的经期跟踪器应用的隐私性，我们不希望应用发出不必要的网络请求。
+
+鉴于我们的 PWA 只由单一的页面构成，对于页面导航请求，我们返回至 `index.html` 主页面，因为没有其他页面，我们也不想请求总是发送到服务器。如果 Fetch API 的 [`Request`](/zh-CN/docs/Web/API/Request) 的只读属性 [`mode`](/zh-CN/docs/Web/API/Request/mode) 值为 `navigate`，意味着它在查找一个 Web 页面，我们使用 FetchEvent 的 [`respondWith()`](/zh-CN/docs/Web/API/FetchEvent/respondWith) 方法来阻止浏览器默认的 fetch 处理，使用 [`caches.match()`](/zh-CN/docs/Web/API/CacheStorage/match) 方法来提供我们自己的响应 Promise。
+
+对于其他所有的请求模式，我们则按照在 [install 事件响应](#在_PWA_安装过程中保存缓存) 里的那样打开缓存，向同一个的 `match()` 方法传递事件请求。它会检查请求是否是已存储的 {{domxref("Response")}} 的键。如果是，我们返回缓存的响应，如果不是，我们返回一个 [404 状态码](/zh-CN/docs/Web/HTTP/Status/404) 作为响应。
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:199]:
```suggestion
对于其他所有的请求模式，我们则按照在 [install 事件响应](#在_PWA_安装过程中保存缓存)里的那样打开缓存，向同样的 `match()` 方法传递事件请求。它会检查请求是否是已存储的 {{domxref("Response")}} 的键。如果是，我们返回缓存的响应，如果不是，我们返回一个 [404 状态码](/zh-CN/docs/Web/HTTP/Status/404)作为响应。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+      } else {
+        // 如果资源不在缓存中，返回 404
+        return new Response(null, { status: 404 });
+      }
+    })()
+  );
+});
+```
+
+当更新 service worker 本身时，不需要更新 VERSION 常量，因为 service worker 脚本内容的任何更改都会触发浏览器去安装新的 service worker。但不论如何，更新版本号是方便开发人员的好习惯，也包括你自己。可以在浏览器中通过[在应用程序工具中检查缓存的名称](#with_developer_tools)（或在源代码工具）来查看当前正在运行的是哪一个版本的 service worker。
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:301]:
```suggestion
当更新 service worker 本身时，不需要更新 VERSION 常量，因为 service worker 脚本内容的任何更改都会触发浏览器去安装新的 service worker。但不论如何，更新版本号是方便开发人员（也包括你自己）的好习惯。可以在浏览器中通过[在应用程序工具中检查缓存的名称](#使用开发人员工具)（或在源代码工具）来查看当前正在运行的是哪一个版本的 service worker。
```

Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:301]:
```suggestion
当更新 service worker 本身时，不需要更新 VERSION 常量，因为 service worker 脚本内容的任何更改都会触发浏览器去安装新的 service worker。但不论如何，更新版本号是方便开发人员（也包括你自己）的好习惯。可以在浏览器中通过[在应用程序工具中检查缓存的名称](#使用开发人员工具)（或在源代码工具）来查看当前正在运行的是哪一个版本的 service worker。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+slug: Web/Progressive_web_apps/Tutorials/CycleTracker/Service_workers
+---
+
+{{PWASidebar}}
+
+{{PreviousMenu("Web/Progressive_web_apps/Tutorials/CycleTracker/Manifest_file", "Web/Progressive_web_apps/Tutorials/CycleTracker")}}
+
+目前为止，我们已经为经期跟踪器编写了 HTML、CSS 和 JavaScript，添加了清单文件用于定义颜色、图标、URL 以及其他的应用特性。我们已经有了一个能运行的 Web 应用！但它还不是一个 PWA。在这一章节，我们将要编写将我们的功能完整、能够作为独立应用分发并且能无缝地在离线状态下运行的 Web 应用转化为 PWA 所需的 JavaScript 脚本。
+
+如果你还没有完成到这一步，请复制这些 [HTML](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/index.html)、[CSS](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/style.css)、[JavaScript](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/app.js)，还有[清单](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/cycletracker.json) JSON 文件的内容，并分别按照以下文件名保存：`index.html`、`styles.css`、`app.js`，以及 `cycletracker.json`。
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:13]:
```suggestion
如果你还没有完成到这一步，请复制这些 [HTML](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/index.html)、[CSS](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/style.css)、[JavaScript](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/app.js)，还有[清单](https://github.com/mdn/pwa-examples/tree/master/cycletracker/manifest_file/cycletracker.json) JSON 文件的内容，并分别按照以下文件名保存：`index.html`、`styles.css`、`app.js` 以及 `cycletracker.json`。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+
+### 使用开发人员工具
+
+你可能会不想每次保存都要更新版本号。在你准备好将你的 PWA 的新版本投入生产并为所有人提供你的 PWA 的新版本前，你可以用注销 service worker 的方法来代替在每次保存时更改版本号。
+
+你可以通过在[浏览器开发人员工具](/zh-CN/docs/Learn/Common_questions/Tools_and_setup/What_are_browser_developer_tools)中点击 `unregister` 按钮来注销一个 service worker。强刷新页面将会重新注册 service worker 并建立新的缓存。
+
+![带有停止和注销 service worker 按钮的 Firefox 开发人员工具应用程序面板](firefox_sw.jpg)
+
+在某些开发者工具里，你可以手动注销一个 service worker，或者你可以选择 service worker 的“重新加载时更新”选项来设置当开发人员工具打开时，每次重新加载都重置并重新激活 service worker。还有一个选项可以绕过 service worker 从网络加载资源。这个面板包含了本教程中我们未覆盖到的特性，但会对你在创建包含 [同步](/zh-CN/docs/Web/Progressive_web_apps/Guides/Offline_and_background_operation#周期性后台同步)和[推送消息](/zh-CN/docs/Web/Progressive_web_apps/Guides/Offline_and_background_operation#推送消息)等特性的更高级的 PWA 时有所帮助，这些都涵盖在了[离线和后台操作](/zh-CN/docs/Web/Progressive_web_apps/Guides/Offline_and_background_operation)中。
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:385]:
```suggestion
在某些开发者工具里，你可以手动注销一个 service worker，或者你可以选择 service worker 的“重新加载时更新”选项来设置当开发人员工具打开时，每次重新加载都重置并重新激活 service worker。还有一个选项可以绕过 service worker 从网络加载资源。这个面板包含了本教程中我们未涵盖的特性，但会对你在创建包含[同步](/zh-CN/docs/Web/Progressive_web_apps/Guides/Offline_and_background_operation#周期性后台同步)和[推送消息](/zh-CN/docs/Web/Progressive_web_apps/Guides/Offline_and_background_operation#推送消息)等特性的更高级的 PWA 时有所帮助，这些内容都涵盖在了[离线和后台操作](/zh-CN/docs/Web/Progressive_web_apps/Guides/Offline_and_background_operation)中。
```

**Comment by yin1999** [comment]
Hey @BSPR0002. I've marked this PR as draft. Please click the "Ready for review" button when you think this PR is ready to merge :). I'll review this PR after that

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+
+我们通过响应以下三种 service worker 事件来实现这些任务：
+
+- [`fetch`](/zh-CN/docs/Web/API/ServiceWorkerGlobalScope/fetch_event) 事件。
+- [`install`](/zh-CN/docs/Web/API/ServiceWorkerGlobalScope/install_event) 事件。
+- [`activate`](/zh-CN/docs/Web/API/ServiceWorkerGlobalScope/activate_event) 事件。
+
+### 版本号
+
+PWA 一但在用户的机器上被安装，唯一通知浏览器有要检索的更新文件的方法就是对 service worker 作出修改。如果 PWA 的其他资源发生了改动——比如 HTML 更新了、CSS 里的 bug 被修复了、`app.js` 添加了新函数、一张图片被压缩了以减少文件体积，等等——你安装的 PWA 的 service worker 是不会知道它需要下载更新资源的。只有 service worker 被改动了，PWA 才会知道它也许该更新缓存了，然后去发起更新的任务。
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:41]:
```suggestion
PWA 一但在用户的机器上被安装，唯一通知浏览器有要检索的更新文件的方法就是对 service worker 作出修改。如果 PWA 的其他资源发生了改动——比如 HTML 更新了、CSS 里的 bug 被修复了、`app.js` 添加了新功能、一张图片被压缩了以减少文件体积，等等——你安装的 PWA 的 service worker 是不会知道它需要下载更新资源的。只有 service worker 被改动了，PWA 才会知道它也许该更新缓存了，然后去发起更新的任务。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+        return new Response(null, { status: 404 });
+      }
+    })()
+  );
+});
+```
+
+当更新 service worker 本身时，不需要更新 VERSION 常量，因为 service worker 脚本内容的任何更改都会触发浏览器去安装新的 service worker。但不论如何，更新版本号是方便开发人员（也包括你自己）的好习惯。可以在浏览器中通过[在应用程序工具中检查缓存的名称](#使用开发人员工具)（或在源代码工具）来查看当前正在运行的是哪一个版本的 service worker。
+
+**备注：** 在对应用的任何资源做出更改后更新版本号是非常重要的，包括 CSS、HTML、JS 代码和图像素材。版本号或者对 service worker 文件所做的任何更改是唯一为你的用户强制更新应用的方式。
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:303]:
```suggestion
**备注：** 在对应用的任何资源（包括 CSS、HTML、JS 代码和图像素材）做出更改后更新版本号是非常重要的。版本号或者对 service worker 文件所做的任何更改是唯一为你的用户强制更新应用的方式。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+        return new Response(null, { status: 404 });
+      }
+    })()
+  );
+});
+```
+
+## 完善 service worker 文件
+
+你的 `sw.js` 文件现在应该和下面的 JavaScript 脚本相似。要注意，当 `APP_STATIC_RESOURCES` 数组中的资源列表发生了任何改动，我们唯一需要在这个 service worker 中更新的常量或函数就是 `VERSION` 的值。
```
Comment (BSPR0002) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:231]:
```suggestion
你的 `sw.js` 文件现在应该和下面的 JavaScript 脚本相似。要注意，当 `APP_STATIC_RESOURCES` 数组中的资源列表发生了任何改动，我们唯一需要在这个 service worker 中更新的常量或功能就是 `VERSION` 的值。
```

**Comment by BSPR0002** [comment]
@yin1999 I have done.

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+
+将上述 `activate` 事件监听器添加至你的 `sw.js` 文件。
+
+### fetch 事件
+
+我们可以利用 [`fetch`](/zh-CN/docs/Web/API/ServiceWorkerGlobalScope/fetch_event) 事件的优势，在用户处于在线状态时阻止已安装的 PWA 发起请求。监听 fetch 事件使得拦截所有请求并使用缓存的响应内容进行响应而不流经网络变得可能。大多数的应用不需要这样的行为。事实上，许多商业模式出于跟踪和营销目的都希望用户定期发送服务器请求。所以，尽管拦截请求对于某些应用来说是一种反面模式，但为了提高我们的经期跟踪器应用的隐私性，我们不希望应用发出不必要的网络请求。
+
+鉴于我们的 PWA 只由单一的页面构成，对于页面导航请求，我们返回至 `index.html` 主页面，因为没有其他页面，我们也不想请求总是发送到服务器。如果 Fetch API 的 [`Request`](/zh-CN/docs/Web/API/Request) 的只读属性 [`mode`](/zh-CN/docs/Web/API/Request/mode) 值为 `navigate`，意味着它在查找一个 Web 页面，我们使用 FetchEvent 的 [`respondWith()`](/zh-CN/docs/Web/API/FetchEvent/respondWith) 方法来阻止浏览器默认的 fetch 处理，使用 [`caches.match()`](/zh-CN/docs/Web/API/CacheStorage/match) 方法来提供我们自己的响应 Promise。
+
+对于其他所有的请求模式，我们则按照在 [install 事件响应](#在_PWA_安装过程中保存缓存)里的那样打开缓存，向同样的 `match()` 方法传递事件请求。它会检查请求是否是已存储的 {{domxref("Response")}} 的键。如果是，我们返回缓存的响应，如果不是，我们返回一个 [404 状态码](/zh-CN/docs/Web/HTTP/Status/404)作为响应。
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:199]:
```suggestion
对于其他所有的请求模式，我们则按照在 [install 事件响应](#在_pwa_安装过程中保存缓存)里的那样打开缓存，向同样的 `match()` 方法传递事件请求。它会检查请求是否是已存储的 {{domxref("Response")}} 的键。如果是，我们返回缓存的响应，如果不是，我们返回一个 [404 状态码](/zh-CN/docs/Web/HTTP/Status/404)作为响应。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+
+我们通过响应以下三种 service worker 事件来实现这些任务：
+
+- [`fetch`](/zh-CN/docs/Web/API/ServiceWorkerGlobalScope/fetch_event) 事件。
+- [`install`](/zh-CN/docs/Web/API/ServiceWorkerGlobalScope/install_event) 事件。
+- [`activate`](/zh-CN/docs/Web/API/ServiceWorkerGlobalScope/activate_event) 事件。
+
+### 版本号
+
+PWA 一但在用户的机器上被安装，唯一通知浏览器有要检索的更新文件的方法就是对 service worker 作出修改。如果 PWA 的其他资源发生了改动——比如 HTML 更新了、CSS 里的 bug 被修复了、`app.js` 添加了新功能、一张图片被压缩了以减少文件体积，等等——你安装的 PWA 的 service worker 是不会知道它需要下载更新资源的。只有 service worker 被改动了，PWA 才会知道它也许该更新缓存了，然后去发起更新的任务。
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:41]:
```suggestion
PWA 一但在用户的机器上被安装，唯一通知浏览器有要检索的更新文件的方法就是对 service worker 作出修改。如果 PWA 的其他资源发生了改动——比如 HTML 更新了、CSS 里的 bug 被修复了、`app.js` 添加了新函数、一张图片被压缩了以减少文件体积，等等——你安装的 PWA 的 service worker 是不会知道它需要下载更新资源的。只有 service worker 被改动了，PWA 才会知道它也许该更新缓存了，然后去发起更新的任务。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 15]
```diff
@@ -0,0 +1,401 @@
...
+
+我们有了版本号，也有了需要缓存的文件。在缓存文件之前，我们需要创建一个在存储应用的静态资源时需要使用的缓存名称。这个缓存名称应当被版本化，以确保在应用更新时可以创建新的缓存同时删除旧的缓存。
+
+#### 任务
+
+使用 `VERSION` 版本号来创建一个版本化的 `CACHE_NAME` 缓存名，作为常量添加到 `sw.js`。
+
+#### 示例方案
+
+我们将缓存命名为 `period-tracker-` 后跟 `VERSION`版本号。鉴于我们的常量声明都是单行的，为了方便阅读，我们将其放在了资源数组常量的前面。
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:106]:
```suggestion
我们将缓存命名为 `period-tracker-` 后跟 `VERSION` 版本号。鉴于我们的常量声明都是单行的，为了方便阅读，我们将其放在了资源数组常量的前面。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 16]
```diff
@@ -0,0 +1,401 @@
...
+
+不要把 service worker 文件包含到缓存资源列表里。
+
+#### 任务
+
+向 `sw.js` 添加用于经期跟踪器 PWA 的缓存资源列表。
+
+#### 示例方案
+
+我们收纳了在本教程的其他章节中所创建的经期跟踪器在离线状态下运作所需的静态资源。我们的 `sw.js` 文件目前是这样的：
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:79]:
```suggestion
我们包括了在本教程的其他章节中所创建的经期跟踪器在离线状态下运作所需的静态资源。我们的 `sw.js` 文件目前是这样的：
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+  "/",
+  "/index.html",
+  "/styles.css",
+  "/app.js",
+  "/cycletrack.json",
+  "/icons/wheel.svg",
+];
+```
+
+我们收纳了 `wheel.svg` 图标，以备你要增强 PWA 的 UI 时使用，比如在没有经期数据时显示 logo，虽然我们的应用目前并没有使用它。
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:94]:
```suggestion
我们包括了 `wheel.svg` 图标，以备你要增强 PWA 的 UI 时使用，比如在没有经期数据时显示 logo，虽然我们的应用目前并没有使用它。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+const CACHE_NAME = `period-tracker-${VERSION}`;
+
+const APP_STATIC_RESOURCES = [ ... ];
+```
+
+我们成功声明了我们的常量：一个唯一的标识符、数组形式的离线资源列表以及随版本号更新的应用缓存名称。现在，让我们把注意力转向安装、更新以及删除无用的缓存资源。
+
+### 在 PWA 安装过程中保存缓存
+
+当用户安装 PWA 或者只是单纯的访问带有 service worker 的网站时，会在 service worker 的作用域触发一个 `install` 事件。我们想要监听这个事件，在安装期间用 PWA 的静态资源充实缓存。每当 service worker 的版本更新，浏览器都会安装新的 service worker 并触发安装事件。
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:119]:
```suggestion
当用户安装 PWA 或者只是单纯的访问带有 service worker 的网站时，会在 service worker 的作用域触发一个 `install` 事件。我们想要监听这个事件，在安装期间用 PWA 的静态资源填充缓存。每当 service worker 的版本更新，浏览器都会安装新的 service worker 并触发安装事件。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 19]
```diff
@@ -0,0 +1,401 @@
...
+    })()
+  );
+});
+```
+
+#### 任务
+
+添加一个安装事件监听器，用于检索并存储 `APP_STATIC_RESOURCES` 列表中的文件到名为 `CACHE_NAME` 的缓存中。
+
+#### 实例方案
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:146]:
```suggestion
#### 示例方案
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 20]
```diff
@@ -0,0 +1,401 @@
...
+        return new Response(null, { status: 404 });
+      }
+    })()
+  );
+});
+```
+
+当更新 service worker 本身时，不需要更新 VERSION 常量，因为 service worker 脚本内容的任何更改都会触发浏览器去安装新的 service worker。但不论如何，更新版本号是方便开发人员（也包括你自己）的好习惯。可以在浏览器中通过[在应用程序工具中检查缓存的名称](#使用开发人员工具)（或在源代码工具）来查看当前正在运行的是哪一个版本的 service worker。
+
+**备注：** 在对应用的任何资源（包括 CSS、HTML、JS 代码和图像素材）做出更改后更新版本号是非常重要的。版本号或者对 service worker 文件所做的任何更改是唯一为你的用户强制更新应用的方式。
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:303]:
```suggestion
**备注**：在对应用的任何资源（包括 CSS、HTML、JS 代码和图像素材）做出更改后更新版本号是非常重要的。版本号或者对 service worker 文件所做的任何更改是唯一为你的用户强制更新应用的方式。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 21]
```diff
@@ -0,0 +1,401 @@
...
+```html
+<!-- 注册应用的 service worker -->
+<script>
+  if ("serviceWorker" in navigator) {
+    navigator.serviceWorker.register("sw.js");
+  }
+</script>
+```
+
+你可以体验功能完整的[CycleTracker 月经周期跟踪 Web 应用](https://mdn.github.io/pwa-examples/cycletracker/service_workers)，并可以在 GitHub 上查看该 [Web 应用的源代码](https://github.com/mdn/pwa-examples/tree/master/cycletracker/service_workers)。是的，它能运行，并且现在正式成为了一个 PWA！
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:362]:
```suggestion
你可以体验功能完整的 [CycleTracker 月经周期跟踪 Web 应用](https://mdn.github.io/pwa-examples/cycletracker/service_workers)，并可以在 GitHub 上查看该 [Web 应用的源代码](https://github.com/mdn/pwa-examples/tree/master/cycletracker/service_workers)。是的，它能运行，并且现在正式成为了一个 PWA！
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 22]
```diff
@@ -0,0 +1,401 @@
...
+### 通过更新版本号并进行强重置
+
+要获得新的缓存，你可以更改[版本号](#版本号)然后进行一次浏览器强刷新。进行强刷新的方式取决于你的浏览器和操作系统：
+
+- Windows：Ctrl+F5，Shift+F5 或 Ctrl+Shift+R。
+- MacOS：Shift+Command+R。
+- MacOS 上的 Safari：按 Option+Command+E 清空缓存，然后按 Option+Command+R。
+- 移动设备：前往浏览器（Android）或操作系统（Samsung、iOS）设置，在高级设置下找到浏览器（iOS）或站点数据（Android、Samsung）网站设置，然后删除经期跟踪器的数据，再重新加载页面。
+
+### 使用开发人员工具
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:377]:
```suggestion
### 使用开发者工具
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 23]
```diff
@@ -0,0 +1,401 @@
...
+
+你可能会不想每次保存都要更新版本号。在你准备好将你的 PWA 的新版本投入生产并为所有人提供你的 PWA 的新版本前，你可以用注销 service worker 的方法来代替在每次保存时更改版本号。
+
+你可以通过在[浏览器开发人员工具](/zh-CN/docs/Learn/Common_questions/Tools_and_setup/What_are_browser_developer_tools)中点击 `unregister` 按钮来注销一个 service worker。强刷新页面将会重新注册 service worker 并建立新的缓存。
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:381]:
```suggestion
你可以通过在[浏览器开发者工具](/zh-CN/docs/Learn/Common_questions/Tools_and_setup/What_are_browser_developer_tools)中点击 `unregister` 按钮来注销一个 service worker。强刷新页面将会重新注册 service worker 并建立新的缓存。
```

File: files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md [hunk 24]
```diff
@@ -0,0 +1,401 @@
...
+
+![带有停止和注销 service worker 按钮的 Firefox 开发人员工具应用程序面板](firefox_sw.jpg)
+
+在某些开发者工具里，你可以手动注销一个 service worker，或者你可以选择 service worker 的“重新加载时更新”选项来设置当开发人员工具打开时，每次重新加载都重置并重新激活 service worker。还有一个选项可以绕过 service worker 从网络加载资源。这个面板包含了本教程中我们未涵盖的特性，但会对你在创建包含[同步](/zh-CN/docs/Web/Progressive_web_apps/Guides/Offline_and_background_operation#周期性后台同步)和[推送消息](/zh-CN/docs/Web/Progressive_web_apps/Guides/Offline_and_background_operation#推送消息)等特性的更高级的 PWA 时有所帮助，这些内容都涵盖在了[离线和后台操作](/zh-CN/docs/Web/Progressive_web_apps/Guides/Offline_and_background_operation)中。
+
+![Edge 开发人员工具显示针对 service worker 的应用程序面板集](edge_sw.jpg)
```
Comment (yin1999) [inline files/zh-cn/web/progressive_web_apps/tutorials/cycletracker/service_workers/index.md:387]:
```suggestion
![Edge 开发者工具显示针对 service worker 的应用程序面板集](edge_sw.jpg)
```

**Review by yin1999 — APPROVED** [review]
Thanks @BSPR0002

**Pull request merged**